### ISSUE TITLE
Graphics-only queue families

### ISSUE THREAD

**Issue by baldurk** [KhronosGroup/Vulkan-Docs#287]
This issue is just a theoretical question for clarification, to avoid armchair spec-lawyering :).

There was some discussion about whether it would be valid to have a graphics-only queue family, that supports graphics & transfer operations but *not* compute.

The relevant sections of the spec that I can see are:

>If an implementation exposes any queue family that supports graphics operations, at least one queue family of at least one physical device exposed by the implementation must support both graphics and compute operations.

and

> All commands that are allowed on a queue that supports transfer operations are also allowed on a queue that supports either graphics or compute operations thus if the capabilities of a queue family include `VK_QUEUE_GRAPHICS_BIT` or `VK_QUEUE_COMPUTE_BIT` then reporting the `VK_QUEUE_TRANSFER_BIT` capability separately for that queue family is optional.

Both in section 4.1 under `VkQueueFamilyProperties`.

To me this reads that graphics `VK_QUEUE_GRAPHICS_BIT` or `VK_QUEUE_COMPUTE_BIT` both imply `VK_QUEUE_TRANSFER_BIT`, such that it doesn't even necessarily have to be reported. However the only implication is that if there exists a queue family with `VK_QUEUE_GRAPHICS_BIT` that there must be at least one with `VK_QUEUE_GRAPHICS_BIT|VK_QUEUE_COMPUTE_BIT`. That might be the same family, or it might not, so graphics capability does not imply compute capability.

In other words is the following configuration valid:

```
Queue family 0: VK_QUEUE_GRAPHICS_BIT
Queue family 1: VK_QUEUE_GRAPHICS_BIT|VK_QUEUE_COMPUTE_BIT
Queue family 2: VK_QUEUE_TRANSFER_BIT
```

In particular, such that queue family 0 does *not* support compute operations? Or is the intention that queue support should be a superset: Graphics queue families imply compute and transfer, compute queue families imply just transfer.

**Comment by krOoze** [comment]
But, but, we love armchair spec-lawyering here :yum: 

Seems obviously valid to me... You can even have multiple families of exactly the same type.

Reporting TRANSFER in that case is truly **optional** (notice: not even **should**):
> [...] thus if the capabilities of a queue family include VK_QUEUE_GRAPHICS_BIT or VK_QUEUE_COMPUTE_BIT then reporting the VK_QUEUE_TRANSFER_BIT capability separately for that queue family is **optional**.

I think if TRANSFER is reported, that the GPU has some specialized HW for that purpose (and if it isn't, then perhaps e.g. the copy will be made using regular compute/graphics pipeline under the hood). But it is up to the vendor to report what will be more obvious to the user.

**Comment by janharaldfredriksen-arm** [comment]
You are both correct.

It's perhaps unfortunate that the wording for this is so complicated. But we wanted to allow the option of exposing compute-only devices while also guaranteeing that compute shaders were available for graphics.

We could perhaps have added a requirement saying that if graphics is supported, then queue family 0 must support graphics+compute, but we didn't.

In summary:
- the configuration in @baldurk's comment is valid
- you can have queue families that only support COMPUTE, only support GRAPHICS, or only support TRANSFER
- transfer operations are supported on COMPUTE and/or GRAPHICS queue, and therefore reporting the TRANSFER capability for such queue families is not necessary, but allowed (i.e., "optional")
 - as @krOoze says, it's up to each vendor to decide whether or not the TRANSFER bit is reported in this case

If that clears things up, could we close this issue?

**Comment by janharaldfredriksen-arm** [comment]
Closing. (See previous comment.)

**Comment by Anteru** [comment]
"guaranteeing that compute shaders were available for graphics" - doesn't this imply that the GRAPHICS bit implicitly includes the COMPUTE bit, hence making graphics a super-set of compute? I get a COMPUTE only and a TRANSFER only queue, but a GRAPHICS queue without COMPUTE and TRANSFER doesn't seem to make much sense. I'd certainly expect exactly what you said: Compute shaders are available for graphics, but a graphics queue is not guaranteed.

**Comment by janharaldfredriksen-arm** [comment]
> doesn't this imply that the GRAPHICS bit implicitly includes the COMPUTE bit,
I didn't mean to imply that.

Implementations are allowed to support graphics-only queues. But, per the first spec quote in Baldur's comment above, we guarantee that if graphics is supported at all, then there is at least one queue that supports graphics and compute.

**Comment by baldurk** [comment]
I guess the question is about what value to place on different things:

* How important is it to actually be able to support graphics-but-not-compute queue families? does anyone actually do this presently (not to my knowledge at least) or plan to in future? If not, making graphics a superset of compute which is in turn a superset of transfer would in my view greatly simplify application use since they can assume things about queue family capabilities.
* How important is free ordering of queue families? Memory types have a number of restrictions on them, both in terms of legal combinations and what order they can come in. I would argue that requiring more capable queue families to be defined before less capable queues would also simplify application use (not as much, but there's value there to end-users). Likewise, I think every implementation I've seen already satisfies this.
* In light of the above, would any IHVs have problems/object if the restrictions on queue families were increased in the spec by adding one or both of these stipulations? If there is an objection to changing 1.0 as is then perhaps this could still be considered for Vulkan 1.1.

I'd argue that while theoretically the spec doesn't have any of these promises/restrictions, in practice everyone already complies with them and is quite unlikely to take advantage of the oddball configurations the spec currently allows.

**Comment by janharaldfredriksen-arm** [comment]
We had quite a lot of discussion around required queue families for 1.0, and what's in the spec is the strongest we could get agreement on.

I'm not aware of free ordering being important to anyone. We may be able to say that if there is a combined graphics and compute queue family, it must be at index 0.
